src/topology/algebra/constructions.lean

```diff
@@ -115,4 +115,40 @@ by simp only [inducing_embed_product.continuous_iff, embed_product_apply, (∘),
 @[to_additive] lemma continuous_coe_inv : continuous (λ u, ↑u⁻¹ : Mˣ → M) :=
 (units.continuous_iff.1 continuous_id).2
 
+/-- The inverse map `units.inv` is continuous with respect to the topology induced by
+  `units.coe_hom` on units. -/
+@[continuity]
+lemma induced_top_cont_inv {X : Type*} [topological_space X] [discrete_topology X] [monoid X] :
+  @continuous _ _ (topological_space.induced (units.coe_hom X) infer_instance) _
+  (units.inv : Xˣ → X) :=
+begin
+  convert continuous_of_discrete_topology,
+  refine @inducing.discrete_topology Xˣ X (topological_space.induced
+    (units.coe_hom X) infer_instance) _ _ (units.coe_hom _)
+    (λ a b h, units.eq_iff.1 h) rfl,
+-- Alternate proof :
+--  refine @embedding.discrete_topology (zmod n)ˣ (zmod n) (topological_space.induced
+--    (units.coe_hom (zmod n)) infer_instance) _ _ (units.coe_hom _) _,
+--  constructor,
+--  { constructor, refl, },
+--  { refine λ a b h, units.eq_iff.1 h, }, }
+end
+
+lemma discrete_topology_of_discrete {X : Type*} [_root_.topological_space X] [discrete_topology X]
+  [monoid X] : discrete_topology Xˣ :=
+begin
+  convert embedding.discrete_topology units.embedding_embed_product,
+  apply @prod.discrete_topology _ _ infer_instance infer_instance infer_instance infer_instance,
+  swap, { refine embedding.discrete_topology (embedding.mk ⟨rfl⟩ mul_opposite.unop_injective), },
+  { apply_instance, },
+end
+
+lemma discrete_prod_units {X Y : Type*} [monoid X] [monoid Y] [topological_space X]
+  [discrete_topology X] [topological_space Y] [discrete_topology Y] : discrete_topology (X × Y)ˣ :=
+begin
+  apply @embedding.discrete_topology _ _ _ _ prod.discrete_topology (units.embed_product _)
+    (units.embedding_embed_product),
+  { apply_instance, },
+  { refine inducing.discrete_topology (mul_opposite.unop_injective) rfl, },
+end
```

Comment:
I think this works, no?
```suggestion
by apply_instance
```

Reply:
No, it does not. The topology is not canonical.

Reply:
I'm not sure I understand. The canonical topology on `(X × Y)ˣ` is the one coming from the embedding into `(X × Y) × (X × Y)`. If `X`, `Y` have the discrete topology, then so does `(X × Y)ˣ`. This is exactly what you're proving here. What's your non-canonical topology doing if it agrees with the canonical one in your use case?

Reply:
The embedding is from `(X × Y)ˣ` to `(X × Y) × (X × Y)ᵐᵒᵖ` . I don't think `(X × Y)ᵐᵒᵖ` has the discrete topology as an `instance`

Reply:
Ah then can you add it? That seems like an obvious omission.

Reply:
Done

Reply:
Does `applyi_instance` now work?

Reply:
Yes

Reply:
Then can you remove it entirely?

Reply:
I am now getting an error in `src/topology/algebra/group/basic.lean` on line 1338 :
invalid 'open' command, unknown declaration 'units.mul_opposite.continuous_op'

however, I did not touch any of that code..

Reply:
`embedding.discrete_topology` is not known to the typeclass system. This works:
```lean
instance foo {X : Type*} [_root_.topological_space X] [discrete_topology X] :
  discrete_topology Xᵐᵒᵖ := sorry

lemma discrete_topology_of_discrete {X : Type*} [_root_.topological_space X] [discrete_topology X]
  [monoid X] : discrete_topology Xˣ :=
begin
  apply embedding.discrete_topology units.embedding_embed_product,
  apply_instance,
end
```